### PR TITLE
Add dynamic channels

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: oak/server/channel.h [hunk 1]
```diff
@@ -23,123 +23,157 @@
...
 
-  // BlockingRead behaves like Read but blocks until a message is available.
-  virtual ReadResult BlockingRead(uint32_t size) { return Read(size); }
+  // Create a message channel and return references to both halves of it.
+  static ChannelHalves Create();
 
-  // Indicate whether a Read operation would return a message.
-  virtual bool CanRead() { return false; }
+ private:
+  friend class MessageChannelReadHalf;
```
Comment (anghelcovici) [inline oak/server/channel.h]: This is pretty much saying that only these halves can use the Message. Is that correct?

File: oak/server/channel.h [hunk 2]
```diff
@@ -23,123 +23,157 @@
...
+        return absl::make_unique<ChannelHalf>(read->Clone());
+        break;
+      case Half::WRITE:
+        return absl::make_unique<ChannelHalf>(write->Clone());
+        break;
+    }
+    return nullptr;
+  }
+
+  enum class Half : int { READ, WRITE };
```
Comment (anghelcovici) [inline oak/server/channel.h]: Why int?

File: oak/server/channel.h [hunk 3]
```diff
@@ -23,123 +23,157 @@
...
+  Half which;
+  std::unique_ptr<MessageChannelReadHalf> read;    // if which==READ
```
Comment (anghelcovici) [inline oak/server/channel.h]: It would be nice to enforce from the type system to have either one or the other. Would absl::variant be an option here?

File: oak/server/channel.h [hunk 4]
```diff
@@ -23,123 +23,157 @@
...
+      : which(Half::WRITE), write(std::move(w)) {}
+
+  std::unique_ptr<ChannelHalf> Clone() {
+    switch (which) {
+      case Half::READ:
+        return absl::make_unique<ChannelHalf>(read->Clone());
+        break;
+      case Half::WRITE:
+        return absl::make_unique<ChannelHalf>(write->Clone());
+        break;
```
Comment (anghelcovici) [inline oak/server/channel.h]: Can we add a default option and print an error message?

File: oak/server/channel.h (hunk 3 above)
Comment (anghelcovici) [reply oak/server/channel.h]: Just seen your comment :)

File: oak/server/logging_node.cc [hunk 5]
```diff
@@ -37,12 +37,13 @@ void LoggingNode::Run() {
       LOG(WARNING) << "Node termination requested";
       return;
     }
-    ReadResult result = channel->Read(INT_MAX);
+    ReadResult result = channel->Read(INT_MAX, INT_MAX);
```
Comment (anghelcovici) [inline oak/server/logging_node.cc]: I know we discussed this, but it still confuses me. Have you thought about having a peek function that returns the size?

File: oak/server/wasm_node.cc [hunk 6]
```diff
@@ -301,9 +352,26 @@ wabt::interp::HostFunc::Callback WasmNode::OakChannelWrite(wabt::interp::Environ
...
+    for (uint32_t ii = 0; ii < handle_count; ii++) {
+      Handle handle = ReadU64(env, handle_offset + (8 * ii));
+      LOG(INFO) << "Transfer channel handle " << handle;
+      ChannelHalf* half = BorrowChannel(handle);
+      if (half == nullptr) {
+        LOG(WARNING) << "Invalid transferred channel handle: " << handle;
+        results[0].set_i32(OakStatus::ERR_BAD_HANDLE);
+        return wabt::interp::Result::Ok;
+      }
+      msg->channels.push_back(half->Clone());
```
Comment (anghelcovici) [inline oak/server/wasm_node.cc]: would emplace_back work here?

File: oak/server/oak_grpc_node.cc [hunk 7]
```diff
@@ -64,8 +64,8 @@ void OakGrpcNode::CompletionQueueLoop() {
 
   Handle req_handle = FindChannel(kGrpcNodeRequestPortName);
   Handle rsp_handle = FindChannel(kGrpcNodeResponsePortName);
-  ChannelHalf* req_half = BorrowChannel(req_handle);
-  ChannelHalf* rsp_half = BorrowChannel(rsp_handle);
+  MessageChannelWriteHalf* req_half = BorrowWriteChannel(req_handle);
```
Comment (anghelcovici) [inline oak/server/oak_grpc_node.cc]: Is there the possibility that the channels is destroyed while being borrowed (and not nullified)? Would it make sense to use a weak_ptr instead?

File: oak/server/channel.h (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/channel.h]: I didn't know `absl::variant` existed, so I've shifted to that.

File: oak/server/channel.h (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/channel.h]: (Overtaken by move to `absl::variant`)

File: oak/server/channel.h (hunk 4 above)
Comment (daviddrysdale) [reply oak/server/channel.h]: (Overtaken by move to `absl::variant`)

File: oak/server/logging_node.cc (hunk 5 above)
Comment (daviddrysdale) [reply oak/server/logging_node.cc]: I don't think a peek function would help, because C++ callers don't need it (they get ownership of the `Message`, so don't have any size concerns) and for Wasm callers (via `wasm::channel_read`) it would expand the syscall surface without adding any new functionality:
 - `wasm::channel_read` will still need to cope with the case that the provided byte/handle buffers are too small, because it can't rely on the node code doing things correctly
 - peeking at sizes can already be done with something like `wasm::channel_read(handle, nil, &size, nil, &count)`.

But you're right that it's a bit hard to follow &ndash; I've had a go at re-writing the `MessageChannelReadHalf::Read` comment.
